openvpn-wire-protocol.xml

```diff
@@ -1542,6 +1545,207 @@ struct datakeys {
+          and the definition of HKDF-Expand-Label in TLS 1.3.
+
+          For OpenVPN the epoch 1 key E_1 are the first 32 the OpenVPN data key. Note that unlike when
+          when using the OpenVPN key directly for encrypting/decryption, E_1 is the same for both direction
+          and basically only the first 32 bytes of key_c2s are used.
+        </t>
+        <t>
+          Note that we start at epoch 1 instead epoch 0 for the key.  This is done to allow for easier
+          implementation for the C-based OpenVPN implementation.
+        </t>
+        <t>
+          For any further key, the key derivation is
+          <artwork> E_N+1 = OVPN-Expand-Label(E_N, "datakey upd", "", 32)</artwork>
+
+
+          The per epoch data channel key is derived via
+          <artwork> K_i = OVPN-Expand-Label(E_i, "epoch key", "", 256) </artwork>
+          
+          The epoch K_i will then be used for the actual data encryption. The encryption key for
+          the AEAD cipher are first bytes from key_c2s and key_s2c respectively. The number of bytes
+          depends on the key size of the cipher.  And the implicit_iv is formed from the
+          first 96 bits (12 bytes) from auth_c2s and auth_s2c respectively.
+        </t>
+        <t>
+          The function OVPN-Expand-Label is identical to the HKDF-Expand-Label function of TLS 1.3 but uses
+          "ovpn " instead of "tls13 " as the label prefix. The hash function for OVPN-Expand-Label is SHA-256.
+
+          <artwork>
+OVPN-Expand-Label(Secret, Label, Context, Length) =
+     HKDF-Expand(Secret, OvpnLabel, Length)
+
+Where OvpnLabel is specified as:
+
+struct {
+    uint16 length = Length;
+    opaque label&lt;6..255> = "ovpn " + Label;
+    opaque context&lt;0..255> = Context;
+} OvpnLabel;
+          </artwork>
+
+          Keep in mind that the &quot; characters, like in TLS 1.3 are part of the label prefix.
+        </t>
+        
+        <t>
+          The key size of 4096 bit or 256 byte is used here to make these epoch data channel keys the same
+          format as the regular data channel keys.
+        </t>
+      </section>
+      <section title="Epoch key usage">
+        <t>
+          When sending packets, a peer will encrypt packet until the limit for the epoch K_i is reach. At this point,
+          the sender will switch to the next epoch key K_i+1. The state for the sending side, only needs to keep
+          the current E_i key and the current encryption key K_i.
+        </t>
+        <t>
+          Each peer is expected to hold the current epoch key K_i, the previous K_i-1 and a number of
+          future keys K_i+1, K_i+2,....,K_i+j. The number of future key j depends on the expected possible
+          data rate of the implementation and needs to be large enough that even when a reasonable amount of
+          packets is lost, the implementation will still be able to authenticate the packets.
```

Comment:
This seems to be an implementation consideration for implementations that are not able to derive future keys on demand from the data channel code (like DCO?). Consider being explicit about that.

And, doesn't the same hold for old keys? If you expect to roll over often, and want to be able to gracefully accept out-of-order packets, you might want to keep more old keys. Or is that not a valid use case?

Reply:
Yeah, this is a tradeoff that I will explicility mention here. Not being able to decrypt an old keys only results a packet loss. Not being able to decrypt a future key results in complete loss of the connection.